### ISSUE TITLE
Upload spec support

### ISSUE THREAD

**Issue by vasilich6107** [comigor/artemis#118]
See #117 for details.

**Comment by vasilich6107** [comment]
I closed the PR cause the file upload spec is not a part of graphql spec and could be implemented in a slightly different way depending on the actual backend.

If you are working with backend based on https://github.com/jaydenseric/graphql-multipart-request-spec here is a solution that works for me.

Define a custom serializer

```
import 'package:http/http.dart';

MultipartFile fromGraphQLUploadToDartMultipartFile(MultipartFile file) => file;
MultipartFile fromDartMultipartFileToGraphQLUpload(MultipartFile file) => file;
```

Update `scalar_maping` section in your `build.yaml`

```
scalar_mapping:
  - graphql_type: Upload
    custom_parser_import: '...location to your custom parser'
    dart_type:
      name: MultipartFile
      imports:
        - 'package:http/http.dart'
```

In your graphql mutation:

```
import 'package:http/http.dart';

final imageFile = await multipartFileFrom(File(path));

final args = Arguments(
        id: id,
        input: Input(file: imageFile));

runMutation(args.toJson());
```

**Comment by vasilich6107** [comment]
After all of that changes you are getting slightly invalid JSON obviously as far as it contains the `MultipartFile` object. I assume this could be acceptable as far as you will not use input classes to encode/decode on the regular basis. The approach I described works perfectly for my case.

If it is not what you need post you own solutions/comments in this issue.

If this is not what you n

**Comment by kateile** [comment]
@vasilich6107 Hi, I trying to follow what you did above but I get this error
```
Invalid argument(s): line 12, column 37 of e:%5C_apps%5Cboko%5Cboko%5Cbuild.yaml: Mapping values are not allowed here. Did you miss a colon earlier?
   ╷
12 │                 custom_parser_import: lib/gql/upload_serializer.dart
   │                                     ^
   ╵
pub finished with exit code 78

```

here is my build.yaml
```yaml
targets:
  $default:
    builders:
      artemis:
        options:
          schema_mapping:
            - output: lib/gql/generated/graphql_api.dart
              schema: lib/gql/schema.graphql
              queries_glob: lib/gql/files/**.graphql
          scalar_mapping:
            - graphql_type: Upload
                custom_parser_import: lib/gql/upload_serializer.dart
                dart_type:
                  name: MultipartFile
                  imports:
                    - 'package:http/http.dart'
```
upload_serializer.dart
```dart
import 'package:http/http.dart';

MultipartFile fromGraphQLUploadToDartMultipartFile(MultipartFile file) => file;
MultipartFile fromDartMultipartFileToGraphQLUpload(MultipartFile file) => file;
```

Schema looks like
```
input DocumentInput {
    "Id as printed in document"
    documentId: String!
    file: Upload!
}

"The `Upload` scalar type represents a file upload."
scalar Upload
```
Earlier I got this error and I was trying to solve it
```
Your `schema` file contains "Upload" scalar, but this scalar is not
configured on `build.yaml`!
Please configure it, following the README on `scalar_mapping`.
```
I think I need to create a parser! Any suggestion?

**Comment by vasilich6107** [comment]
Cause you have wrong indentation in yaml

```
scalar_mapping:
  - graphql_type: Upload
    custom_parser_import: 'package:running_hours/helpers/artemis.dart'
    dart_type:
      name: MultipartFile
      imports:
        - 'package:http/http.dart'
```

**Comment by kateile** [comment]
Thank you. You saved my day

**Comment by ram231** [comment]
> After all of that changes you are getting slightly invalid JSON obviously as far as it contains the `MultipartFile` object. I assume this could be acceptable as far as you will not use input classes to encode/decode on the regular basis. The approach I described works perfectly for my case.
> 
> If it is not what you need post you own solutions/comments in this issue.

@vasilich6107 does the current `multipartFileFrom` function exist in http package? I kept on searching that function but it doesn't exist in the http package.

**Comment by vasilich6107** [comment]
@ram231 it exists in graphql_flutter package
Check the Upload section of graphql_flutter docs

**Comment by ram231** [comment]
Thanks, @vasilich6107. Does the `execute` function work on artemis or should I just use the `mutate` from graphql_flutter? artemis doesn't work on my end so I'll try to use `graphql_flutter`

**Comment by PlugFox** [comment]
You can upload with using REST client in this way:
https://gist.github.com/PlugFox/6934f75b46c9d73951c2d6c1b40415ce

**Comment by vasilich6107** [comment]
@ram231 do not use `Atremis` as grpahlq client. Use `graphql_flutter`

**Comment by PlugFox** [comment]
https://gist.github.com/PlugFox/f2c6da19315754384e487f0634aa063e

graphql extension for support artemis codegen (query and result)

**Comment by darmie** [comment]
@vasilich6107 your example didn't work with json_serializer

```Could not generate `fromJson` code for `videos` because of type `MultipartFile`.```

What do you suggest?

[UPDATE] I have fixed the problem for List<MultipartFile>

**Comment by fikretsengul** [comment]
> @vasilich6107 your example didn't work with json_serializer
> 
> `` Could not generate `fromJson` code for `videos` because of type `MultipartFile`. ``
> 
> What do you suggest?
> 
> [UPDATE] I have fixed the problem for List

How did you manage to fix it? Would you like to share your solution? Thank you.

**Comment by darmie** [comment]
@fikretsengul  I got your message, let me give a break down of what I did. 

1. First of all create a coercers file

> coercers.dart
```dart
// @dart = 2.12

import 'package:http/http.dart' as http;

export 'package:http/http.dart';

DateTime fromGraphQLTimeToDartDateTime(String time) =>
    DateTime.parse(time);
String fromDartDateTimeToGraphQLTime(DateTime date) =>
    date.toUtc().toIso8601String();

DateTime? fromGraphQLTimeToDartDateTimeNullable(String time) =>
    DateTime.parse(time);
String? fromDartDateTimeToGraphQLTimeNullable(DateTime? date) =>
    date?.toUtc().toIso8601String();





http.MultipartFile fromGraphQLUploadToDartMultipartFile(http.MultipartFile u) => u;
http.MultipartFile fromDartMultipartFileToGraphQLUpload(http.MultipartFile file) => file;


List<http.MultipartFile> fromGraphQLListUploadToDartListMultipartFile(List<http.MultipartFile> u) => u;
List<http.MultipartFile> fromDartListMultipartFileToGraphQLListUpload(List<http.MultipartFile> u) => u
```

> build.yaml
```yaml
targets:
  $default:
    builders:
      artemis:
        options:
          schema_mapping:
            ...
          scalar_mapping:
            - custom_parser_import: 'package:project_name/path/to/coercers.dart'
              graphql_type: Time
              dart_type: DateTime
              use_custom_parser: true
            - custom_parser_import:  'package:project_name/path/to/coercers.dart'
              graphql_type: Upload # <-- this is how I represent File type in GraphQL
              dart_type: MultipartFile # <-- it converts to http.MultipartFile
              use_custom_parser: true
```

**Comment by fikretsengul** [comment]
Thank you so much for the attention and fast response. I've checked your solution which I also tried but still no success. My configurations are:

Flutter version: Flutter 2.2.0 • channel stable •
Dart version: Dart SDK version: 2.13.0 (stable)
Flutter environment in pubspec.yaml: sdk: ">=2.12.2 <3.0.0"

I'm using flutter with null-safety.

upload_scalar.dart

```
import 'package:http/http.dart';

MultipartFile fromGraphQLUploadToDartMultipartFile(MultipartFile file) => file;
MultipartFile fromDartMultipartFileToGraphQLUpload(MultipartFile file) => file;
List<MultipartFile> fromGraphQLListUploadToDartListMultipartFile(
        List<MultipartFile> files) =>
    files;
List<MultipartFile> fromDartListMultipartFileToGraphQLListUpload(
        List<MultipartFile> files) =>
    files;
```

build.yaml

```
targets:
  $default:
    builders:
      artemis:
        options:
          schema_mapping:
            - schema: lib/infrastructure/graphql/schema.graphql
              naming_scheme: pathedWithFields
              queries_glob: lib/infrastructure/graphql/queries/**.graphql
              output: lib/infrastructure/graphql/graphql_api.dart
          scalar_mapping:
            - graphql_type: DateTime
              custom_parser_import: "lib/infrastructure/graphql/scalars/datetime_scalar.dart"
              dart_type: DateTime
              use_custom_parser: true
            - graphql_type: Upload
              custom_parser_import: "lib/infrastructure/graphql/scalars/upload_scalar.dart"
              dart_type:
                name: MultipartFile
                imports:
                  - "package:http/http.dart"
              use_custom_parser: true
```

The status of graphql_api.graphql.dart

https://imgur.com/vfeUdNk.png

**Comment by darmie** [comment]
@fikretsengul  I don't think your custom parser is properly set, it should be the import path, not the relative directory... 

change this 
```
custom_parser_import: "lib/infrastructure/graphql/scalars/datetime_scalar.dart"
```

to this, where your package name is the name of your project
```
custom_parser_import: "package:project_name/infrastructure/graphql/scalars/datetime_scalar.dart"
```

**Comment by darmie** [comment]
@fikretsengul you just need to create the coercer for that missing part

add this to  the upload_scalar.dart
```dart
List<MultipartFile> fromGraphQLListUploadToDartListMultipartFileNullable(List<MultipartFile> files) => files;
List<MultipartFile> fromDartListMultipartFileToGraphQLListUploadNullable(List<MultipartFile> files) =>files;
```

**Comment by fikretsengul** [comment]
@darmie I tried that too with flutter clean and re-run the code generator but nothing changes. I don't understand what is wrong with it. Everything seems right about imports but it doesn't recognize those functions I guess.

https://imgur.com/Mvp8xQj.png

You can also check it in details:
https://github.com/comigor/artemis/issues/320#issuecomment-857713994

**Comment by darmie** [comment]
I am not sure your import is right, it's starting with `lib/` shouldn't it be relative to the package?
I also noticed that you replaced everything in your upload_scalar file... can you get on a Google Meet call?
@fikretsengul

**Comment by fikretsengul** [comment]
> I am not sure your import is right, it's starting with `lib/` shouldn't it be relative to the package?
> I also noticed that you replaced everything in your upload_scalar file... can you get on a Google Meet call?
> @fikretsengul

It would be appreciated. Google Meet or Zoom is ok for me :)

**Comment by darmie** [comment]
Cool, check twitter @fikretsengul

**Comment by Khosbayar** [comment]
Hey @darmie  and @fikretsengul , 
I'm facing the same issue as yours, and after I tried @darmie 's approach, I got an error below:
```
This builder requires Dart inputs without syntax errors.
However, package:leialink_v1/graphql/serializer/graphql_serializer.dart (or an existing part) contains the following errors.
graphql_serializer.dart:18:55: This requires the 'non-nullable' language feature to be enabled.
graphql_serializer.dart:18:7: This requires the 'non-nullable' language feature to be enabled.
graphql_serializer.dart:15:9: This requires the 'non-nullable' language feature to be enabled.
```

The problem is in here:
```
DateTime? fromGraphQLTimeToDartDateTimeNullable(String time) =>
    DateTime.parse(time);
String? fromDartDateTimeToGraphQLTimeNullable(DateTime? date) =>
    date?.toUtc().toIso8601String();
```

Any idea how to fix it? Seems like not that complicated one but I don't understand. 😄

**Comment by vasilich6107** [comment]
@darmie 

released new `7.1.0-beta.0`
with updated converters generation

`file: Upload` 
- `fromGraphQLUploadNullableToDartMultipartFileNullable`
- `fromDartMultipartFileNullableToGraphQLUploadNullable`

`file: Upload!`
- `fromGraphQLUploadToDartMultipartFile`
- `fromDartMultipartFileToGraphQLUpload`

`file: [Upload]`
- `fromGraphQLListNullableUploadNullableToDartListNullableMultipartFileNullable`
- `fromDartListNullableMultipartFileNullableToGraphQLListNullableUploadNullable`

`file: [Upload]!`
- `fromGraphQLListUploadNullableToDartListMultipartFileNullable`
- `fromDartListMultipartFileNullableToGraphQLListUploadNullable`

`file: [Upload!]!`
- `fromGraphQLListUploadToDartListMultipartFile`
- `fromDartListMultipartFileToGraphQLListUpload`

**Comment by vasilich6107** [comment]
Added samples for some possible Upload types
https://github.com/artflutter/graphql_toolchain
